components/txn_types/src/types.rs

```diff
@@ -587,6 +587,80 @@ impl WriteBatchFlags {
     }
 }
 
+/// The position info of the last actual write (PUT or DELETE) of a LOCK or
+/// ROLLBACK record.
+#[derive(Clone, Eq, PartialEq)]
+pub enum LastChange {
+    Unknown,
+    Exist(LastChangePosition),
+    NotExist, // Either there is no previous write of the key **or the last write is a DELETE**.
+}
+
+impl LastChange {
+    pub fn make_exist(last_change_ts: TimeStamp, estimated_versions_to_last_change: u64) -> Self {
+        assert!(!last_change_ts.is_zero());
+        assert!(estimated_versions_to_last_change > 0);
+        LastChange::Exist(LastChangePosition {
+            last_change_ts,
+            estimated_versions_to_last_change,
+        })
+    }
+
+    // How `LastChange` is stored.
+
+    pub fn serialize(&self) -> (TimeStamp, u64) {
+        match self {
+            LastChange::Unknown => (TimeStamp::zero(), 0),
+            LastChange::Exist(p) => (p.last_change_ts, p.estimated_versions_to_last_change),
+            LastChange::NotExist => (TimeStamp::zero(), 1),
+        }
+    }
+
+    pub fn deserialize(last_change_ts: TimeStamp, estimated_versions_to_last_change: u64) -> Self {
+        if last_change_ts.is_zero() {
+            if estimated_versions_to_last_change > 0 {
+                LastChange::NotExist
+            } else {
+                LastChange::Unknown
+            }
+        } else {
+            Self::make_exist(last_change_ts, estimated_versions_to_last_change)
+        }
+    }
+}
+
+impl Debug for LastChange {
```

Comment:
Emm I forgot why I chose to impl it manually. Since it's done we can keep it as derived Debug is more verbose.
`last_change: Exist(LastChangePosition { last_change_ts: TimeStamp(80), estimated_versions_to_last_change: 4 })`
`last_change: Exist { last_change_ts: 80, estimated_versions_to_last_change: 4 }`